src/spy-types.ts

```diff
@@ -1,25 +1,61 @@
-import 'jasmine';
+/// <reference types="jasmine" />
+import { Observable } from "rxjs";
 
-/**
- * The Auto Spy Object type
- */
+export type Spy<T> = { [k in keyof T]: AddSpyTypes<T[k]> };
 
-export type Spy<T> = T & {
-  [k in keyof T]: AsyncSpyFunction;
-}
+export type AddSpyTypes<T> = T extends (...args: any[]) => any
+  ? AddSpyByReturnTypes<T>
+  : T;
+
+// this will add spy types on every proeprty of the given type.
+// not only functions by return value.
+// export type AddSpyTypes<T> = T extends (...args: any[]) => any
+//   ? AddSpyOnReturnTypes<T>
+//   : T extends Promise<any>
+//     ? AddSpyOnPromise<T>
+//     : T extends Observable<any> ? AddSpyOnObservable<T> : T;
 
-/**
- * Originally in Jasmine, Spy is a function and SpyObj is an Object
- * But we decided to use "Spy" for objects, and "SpyFunction" to functions.
- */
-export interface AsyncSpyFunction extends jasmine.Spy {
-  (...params: any[]): any;
-  and: AsyncSpyFunctionAnd
+export interface PromiseSpy<T> {
+  resolveWith(value: T): void;
+  rejectWith(value: any): void;
 }
 
-export interface AsyncSpyFunctionAnd extends jasmine.SpyAnd {
-  nextWith(value: any): void;
+export interface ObservableSpy<T> {
+  nextWith(value: T): void;
   nextWithError(value: any): void;
-  resolveWith(value: any): void;
-  rejectWith(value: any): void;
-}
\ No newline at end of file
+}
+
+export type AddSpyOnFunction<T extends (...args: any[]) => R, R> = T & {
+  and: jasmine.Spy;
```

Comment:
The jasmine types actually a bit poor.  they don't use type information that they can get from the function (args/return value.)
We can add our own types for function spy and make the function return type more safe